pkg: ucglib: add support for color graphics library

basilfx opened the pull request:
### Contribution description

This PR adds support for a color graphics library named [Ucglib](https://github.com/olikraus/ucglib). Its created by the same author as U8g2, which we already support.

Interfaces are pretty similar, but Ucglib has only support for SPI displays at this moment. There is no terminal display, but SDL2 is supported.

It's not the fastest library, because it writes directly to the display: [see this video](https://www.youtube.com/watch?v=gpRIfOITF3k) (that's using 10MHz SPI and a 48MHz MCU). I'm using a SSD1331 OLED display, but [many others](https://github.com/olikraus/ucglib/pull/85) are supported too.

Its currently WIP, because I am waiting for an upstream PR to get merged. Until then, it is referring to my own fork.

![img_0346](https://user-images.githubusercontent.com/815976/36565043-8e915078-181f-11e8-8baa-dd8de6bc6c9c.JPG)

### Issues/PRs references

Depends on https://github.com/olikraus/ucglib/pull/85.

kaspar030 commented on pkg/ucglib/patches/0001-add-riot-makefiles.patch:
```diff
@@ -0,0 +1,62 @@
+From 6bdd654ab1b904d245c57765262e1f6736224ddb Mon Sep 17 00:00:00 2001
```
please just add these files to pkg/ucglib, and copy them in some Makefile recipe. No need to carry around patch files...

basilfx replied:
Fixed

aabadie commented:
Very nice one @basilfx !

I see that Ucglib also supports the ILI9341 which is used by one [ST discovery board](http://www.st.com/en/evaluation-tools/32f429idiscovery.html) and I know some people that have one. I may be able to test with that.

kaspar030 commented on tests/pkg_ucglib/logo.h:
```diff
@@ -0,0 +1,1178 @@
+/*
```
any reason this is a header (instead of .c)?

basilfx replied:
Not really. Most converters I tried are called something like `img2header` so I went for a header file.

Is there a difference?

basilfx replied:
Ping @kaspar030 :-)

kaspar030 replied:
no difference.

aabadie commented on pkg/ucglib/patches/0001-add-riot-makefiles.patch:
```diff
@@ -0,0 +1,62 @@
+From 6bdd654ab1b904d245c57765262e1f6736224ddb Mon Sep 17 00:00:00 2001
```
I think this patch could be replaced by Makefiles in the pkg directory that are copied to the ucglib build directory, see semtech-loramac package for an example. It will make future maintenance easier.

basilfx replied:
Fixed.

aabadie commented on tests/pkg_ucglib/Makefile:
```diff
@@ -0,0 +1,44 @@
+TEST_PIN_RESET ?= GPIO_PIN\(0,0\)
+
+ifeq ($(TEST_OUTPUT),3)
+TEST_DISPLAY ?= ucg_dev_ssd1331_18x96x64_univision
```
indentation should be 2 spaces

aabadie commented on tests/pkg_ucglib/Makefile:
```diff
@@ -0,0 +1,44 @@
+
+# features depend on output type
+ifeq ($(TEST_OUTPUT),2)
+USEMODULE += ucglib_sdl
```
same here about indentation

aabadie commented on tests/pkg_ucglib/Makefile:
```diff
@@ -0,0 +1,44 @@
+endif
+
+ifeq ($(TEST_OUTPUT),3)
+FEATURES_REQUIRED += periph_gpio periph_spi
```
and here

basilfx commented:
I'll fix the comments in follow-ups. 

I will also check if I can add a dummy output, so that we can at least assert some aspects of this display.

basilfx commented:
I have fixed the comments and rebased the repository. Also added tests.

Unlike U8g2, this package does not come with a console-based device driver. Therefore, I implemented a dummy driver that just prints a few of the internal commands to screen. Not the best option, but allows for a basic functionality test. Otherwise it would fail Murdock for not having SDL available (that would be the only native-only option).

basilfx commented:
Updated the PR to fix some additional bugs. Murdock seems happy, apart from the `fixup` commit messages.

basilfx commented:
I received a [ST7735](https://www.ebay.com/itm/Colorful-SPI-TFT-LCD-Display-ST7735-128X128-Replace-Nokia-5110-3310-LCD-1-44/282668986643?hash=item41d0627113:g:Cy0AAOSwXXxZOmn-) 1.8" TFT display, and that one also works fine/out of the box. It uses SPI.

![img_4646](https://user-images.githubusercontent.com/815976/37174588-3ee3b1a8-2317-11e8-889e-96fff2526cff.JPG)

(using `TEST_DISPLAY ?= ucg_dev_st7735_18x128x160` and `TEST_DISPLAY_EXT ?= ucg_ext_st7735_18`)

smlng commented:
Nice, I think I have one or two of those as well - will check it out (soonish).

smlng commented on pkg/ucglib/patches/0001-add-RIOT-OS-interface.patch:
```diff
@@ -0,0 +1,248 @@
+ 
+ #ifdef __cplusplus
+ }
+diff --git a/csrc/ucg_riotos.c b/csrc/ucg_riotos.c
```
don't we put such extra files into `contrib`, or some other folder in `pkg/<name>/`, see emb6, lwip, or fatfs for instance.

basilfx replied:
If I would put the files in `contrib` (which would be a new module), I still need to patch `ucg.h` and include the declarations over there (user should only include `ucg.h` IMHO). 

Not sure if that is a preferred option?

smlng replied:
I'm unsure, too - but introducing a completely new file solely for RIOT via a patch looks weird, too. At least put into a separate Patch, so that changes to the original library are self contained and the RIOT adaption (layer), too. That makes adaptions easier, I would say.

basilfx replied:
I could also copy it before compilation, just like the Makefiles?

basilfx replied:
@kaspar030 I think you have an opinion on this?

aabadie replied:
> I could also copy it before compilation, just like the Makefiles?

This is what is done with [tweetnacl](https://github.com/RIOT-OS/RIOT/tree/master/pkg/tweetnacl) package, so it might be the way to go.

basilfx replied:
I have moved the files similar to the `tweetnacl` package.

bergzand commented:
> Just in case, I adapted RIOT to the STM32F429I-DISC1 board and I'm planning to add some code to make this PR work with its ILI9341 screen.

I've got like 90% of pull request ready code for a generic RIOT driver for that display [here](https://github.com/bergzand/RIOT/tree/pr/ili9341) (including a test example)

aabadie commented:
> I've got like 90% of pull request ready code for a generic RIOT driver for that display here (including a test example)

Just had a brief look and you should improve the initialization parameters. Otherwise, it looks good !
My branch with the board port is [here](https://github.com/aabadie/RIOT/tree/pr/boards/stm32f429i-disc1), I'll give it a try but not within the 2 coming weeks.

aabadie commented:
@bergzand, see f93c3c0772ddfcf04bf2c605d01f4a807f8c0ed2 and 944bf14a5151784c55cedca274fd5adaf61ec581 for the small things I changed.

basilfx commented:
Rebased. It still works for me:

![img_7259](https://user-images.githubusercontent.com/815976/38269208-2e7f8756-3780-11e8-83aa-b35ec537289b.JPG)

aabadie commented on tests/pkg_ucglib/main.c:
```diff
@@ -0,0 +1,154 @@
+    puts("Initializing to SPI.");
+
+    ucg_SetPins(&ucg, pins, pins_enabled);
+    ucg_SetDevice(&ucg, TEST_SPI);
```
I think it's better to use `SPI_DEV(TEST_SPI)`

basilfx replied:
Fixed.

aabadie commented on tests/pkg_ucglib/main.c:
```diff
@@ -0,0 +1,154 @@
+
+#include <stdio.h>
+
+#include "periph/gpio.h"
```
If test_output is test_output_spi, then you'll also have to include "periph/spi.h"

basilfx replied:
Fixed.

aabadie commented on tests/pkg_ucglib/main.c:
```diff
@@ -0,0 +1,154 @@
+    while (true) {
+        ucg_ClearScreen(&ucg);
+
+        if (screen == 0) {
```
maybe use a switch/case ?

basilfx replied:
Fixed.

aabadie commented on tests/pkg_ucglib/main.c:
```diff
@@ -0,0 +1,154 @@
+    /* start drawing in a loop */
+    puts("Drawing on screen.");
+
+    while (true) {
```
while (1) ? (minor)

basilfx replied:
Fixed.

smlng commented on tests/pkg_ucglib/main.c:
```diff
@@ -0,0 +1,160 @@
+
+#if TEST_OUTPUT == TEST_OUTPUT_SPI
+/**
+ * @name    RIOT-OS pin maping of Ucglib pin numbers to RIOT-OS GPIO pins.
```
`@name` is only used for doxygen groups (`@{` ... `@}`) so `@brief` would be correct here. Besides the fact that this is  a c file and will not be parsed by doxygen, so you can omit any doxygen tags and use plain C comments.

Applies below too

basilfx replied:
Kept the comment, but changed to `@brief`. I think it's better to adhere to one style of comments :-)

smlng commented on tests/pkg_ucglib/main.c:
```diff
@@ -0,0 +1,160 @@
+    while (1) {
+        ucg_ClearScreen(&ucg);
+
+
```
nit-picky: remove dup empty line

basilfx replied:
Fixed.

smlng commented on tests/pkg_ucglib/logo.h:
```diff
@@ -0,0 +1,1178 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
```
missing doxygen header

basilfx replied:
Added header.

smlng commented on pkg/ucglib/src/csrc/ucg_riotos.c:
```diff
@@ -0,0 +1,161 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
```
missing doxygen header, though its not as import for C files as for header files

basilfx replied:
Added header.
